hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationRegistry.java

```diff
@@ -0,0 +1,365 @@
+
+        Operation op = invocation.op;
+        if (op.getCallId() != 0) {
+            invocations.remove(op.getCallId());
+        }
+
+        invocations.put(callId, invocation);
+        setCallId(invocation.op, callId);
+    }
+
+    /**
+     * Not every call needs to be registered. A call that is local and has no backups, doesn't need to be registered
+     * since there will not be a remote machine sending a response back to the invocation.
+     *
+     * @param invocation
+     * @return true if registration is required, false otherwise.
+     */
+    private boolean skipRegistration(Invocation invocation) {
+        if (invocation.remote) {
+            return false;
+        }
+
+        if (invocation.op instanceof BackupAwareOperation) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Deregisters an invocation.
+     *
+     * If the invocation already already deregisterd or the registration was skipped, the call is ignored.
+     *
+     * @param invocation the Invocation to deregister.
+     */
+    public void deregister(Invocation invocation) {
+        long callId = invocation.op.getCallId();
+
+        // if an invocation skipped registration (so call id is 0) we don't need to deregister it.
+        if (callId == 0) {
+            return;
+        }
+
+        invocations.remove(callId);
+    }
+
+    public Collection<Invocation> invocations() {
+        return invocations.values();
+    }
+
+    /**
+     * Returns the number of pending invocations.
+     *
+     * @return the number of the pending invocations.
+     */
+    public int size() {
+        return invocations.size();
+    }
+
+    /**
+     * Checks if there are any pending invocations.
+     *
+     * @return true if there are any pending invocations, false otherwise.
+     */
+    public boolean isEmpty() {
+        return invocations.isEmpty();
+    }
+
+    /**
+     * Gets the invocation for the given call id.
+     *
+     * @param callId the callId.
+     * @return the found Invocation or null if no invocation was found.
+     */
+    public Invocation get(long callId) {
+        return invocations.get(callId);
+    }
+
+    /**
+     * Notifies the invocation that a Response is available.
+     *
+     * @param response
+     */
+    public void notify(Response response) {
+        if (response instanceof NormalResponse) {
+            notifyNormalResponse((NormalResponse) response);
+        } else if (response instanceof BackupResponse) {
+            notifyBackupComplete(response.getCallId());
+        } else if (response instanceof CallTimeoutResponse) {
+            notifyCallTimeout((CallTimeoutResponse) response);
+        } else if (response instanceof ErrorResponse) {
+            notifyErrorResponse((ErrorResponse) response);
+        } else {
+            logger.severe("Unrecognized response: " + response);
+        }
+    }
+
+    public void notifyBackupComplete(long callId) {
```

Comment:
this is an inconsistency in a method signature. all other notifyXY() methods accept a *Response as an argument.